### PR TITLE
Kafka streaming source enhancements

### PR DESCRIPTION
**Pull request by albertshau**
Added properties to support specifying the partitions to read
from, as well as initial offsets for each partition. Also changed
it to only read from a single topic to simplify configuration.
In addition, added optional fields to allow including the partition
and offset in the output records.

### CODE REVIEW

**Comment by russorat** [comment]
@abhinavbansal246 can you provide an initial review here?

File: hydrator-common/src/main/java/co/cask/hydrator/common/KeyValueListParser.java [hunk 1]
```diff
@@ -67,7 +68,7 @@ private KeyValueIterable(String kvList) {
     private final Iterator<String> pairIter;
 
     private KeyValueIterator(String kvList) {
-      pairIter = Splitter.on(pairDelimiter).split(kvList).iterator();
+      pairIter = Splitter.on(pairDelimiter).trimResults().split(kvList).iterator();
```
Comment (abhinavbansal246) [inline hydrator-common/src/main/java/co/cask/hydrator/common/KeyValueListParser.java]: Why do we need to strip white spaces?

File: spark-plugins/docs/Kafka-streamingsource.md [hunk 2]
```diff
@@ -18,9 +18,22 @@ Properties
...
 
-**topics:** Comma-separated list of Kafka topics to read from. (Macro-enabled)
+**topic:** The Kafka topic to read from. (Macro-enabled)
+
+**partitions:** List of topic partitions to read from. If not specified, all partitions will be read. (Macro-enabled)
+
+**defaultInitialOffset:** The default initial offset for all topic partitions.
+An offset of -2 means the smallest offset. An offset of -1 means the latest offset. Defaults to -1.
+Offsets are inclusive. If an offset of 5 is used, the message at offset 5 will be read.
+If you wish to set different initial offsets for different partitions, use the initialPartitionOffsets property.
```
Comment (abhinavbansal246) [inline spark-plugins/docs/Kafka-streamingsource.md]: add (Macro-enabled)

File: spark-plugins/docs/Kafka-streamingsource.md [hunk 3]
```diff
@@ -18,9 +18,22 @@ Properties
...
+
+**initialPartitionOffsets:** The initial offset for each topic partition. If this is not specified,
+all partitions will use the same initial offset, which is determined by the defaultInitialOffset property.
+Any partitions specified in the partitions property, but not in this property will use the defaultInitialOffset.
+An offset of -2 means the smallest offset. An offset of -1 means the latest offset.
+Offsets are inclusive. If an offset of 5 is used, the message at offset 5 will be read.
```
Comment (abhinavbansal246) [inline spark-plugins/docs/Kafka-streamingsource.md]: add (Macro-enabled)

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java [hunk 4]
```diff
@@ -116,6 +161,16 @@ public String getKeyField() {
   }
 
   @Nullable
+  public String getPartitionField() {
+    return partitionField;
```
Comment (abhinavbansal246) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java]: Do we need separate getter for partitionField?
partitionField has a public access modifier

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java [hunk 5]
```diff
@@ -116,6 +161,16 @@ public String getKeyField() {
...
+  }
+
+  @Nullable
+  public String getOffsetField() {
+    return offsetField;
```
Comment (abhinavbansal246) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java]: same as above

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java [hunk 6]
```diff
@@ -128,31 +183,37 @@ public Schema getSchema() {
...
+        if (fieldType != Schema.Type.BYTES) {
           throw new IllegalArgumentException("The key field must be of type bytes or nullable bytes.");
         }
         keyFieldExists = true;
+      } else if (fieldName.equals(partitionField)) {
+        if (fieldType != Schema.Type.INT) {
+          throw new IllegalArgumentException("The partition field must be of type int.");
+        }
+        partitionFieldExists = true;
+      } else if (fieldName.equals(offsetField)) {
```
Comment (abhinavbansal246) [inline spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java:216]: Do we need to do any check for offsetField? If not can we remove this "else if (fieldName.equals(offsetField))"

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java (hunk 4 above)
Comment (albertshau) [reply spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java]: not sure I understand. All fields in this class are private.

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java (hunk 6 above)
Comment (albertshau) [reply spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java:216]: oops, forgot to add the type check (it must be a long). will add it.

File: spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java (hunk 4 above)
Comment (abhinavbansal246) [reply spark-plugins/src/main/java/co/cask/hydrator/plugin/spark/KafkaConfig.java]: please ignore it, partitionField is private

**Pull request merged**